metrics/ci_integrations/test/source/github_actions/adapter/github_actions_source_client_adapter_test.dart

```diff
@@ -0,0 +1,1022 @@
+      () {
+        whenFetchWorkflowRuns(
+          withArtifactsPage: defaultArtifactsPage,
+          withJobsPage: emptyWorkflowRunJobsPage,
+        ).thenSuccessWith(defaultRunsPage);
+
+        when(githubActionsClientMock.fetchRunJobsNext(emptyWorkflowRunJobsPage))
+            .thenSuccessWith(defaultJobsPage);
+
+        final result = adapter.fetchBuilds(testDataGenerator.jobName);
+
+        expect(result, completion(equals(defaultBuildData)));
+      },
+    );
+
+    test(
+      ".fetchBuilds() fetches builds using pagination for run artifacts",
+      () {
+        whenFetchWorkflowRuns(
+          withArtifactsPage: emptyArtifactsPage,
+          withJobsPage: defaultJobsPage,
+        ).thenSuccessWith(defaultRunsPage);
+
+        when(githubActionsClientMock.fetchRunArtifactsNext(emptyArtifactsPage))
+            .thenSuccessWith(defaultArtifactsPage);
+
+        final result = adapter.fetchBuilds(testDataGenerator.jobName);
+
+        expect(result, completion(equals(defaultBuildData)));
+      },
+    );
+
+    test(
+      ".fetchBuilds() fetches coverage for each build using pagination for run artifacts",
+      () async {
+        whenFetchWorkflowRuns(
+          withArtifactsPage: emptyArtifactsPage,
+          withJobsPage: defaultJobsPage,
+        ).thenSuccessWith(defaultRunsPage);
+
+        when(githubActionsClientMock.fetchRunArtifactsNext(emptyArtifactsPage))
+            .thenSuccessWith(defaultArtifactsPage);
+
+        final expectedCoverage = [
+          testDataGenerator.coverage,
+          testDataGenerator.coverage,
+        ];
+
+        final result = await adapter.fetchBuilds(testDataGenerator.jobName);
+        final actualCoverage =
+            result.map((buildData) => buildData.coverage).toList();
+
+        expect(actualCoverage, equals(expectedCoverage));
+      },
+    );
+
+    test(
+      ".fetchBuilds() throws a StateError if fetching a workflow runs page fails",
+      () {
+        whenFetchWorkflowRuns(
+          withArtifactsPage: defaultArtifactsPage,
+          withJobsPage: defaultJobsPage,
+        ).thenErrorWith();
+
+        final result = adapter.fetchBuilds(testDataGenerator.jobName);
+
+        expect(result, throwsStateError);
+      },
+    );
+
+    test(
+      ".fetchBuilds() throws a StateError if looking for the coverage artifact fails",
+      () {
+        whenFetchWorkflowRuns(
+          withArtifactsPage: defaultArtifactsPage,
+          withJobsPage: defaultJobsPage,
+        ).thenSuccessWith(defaultRunsPage);
+
+        when(githubActionsClientMock.fetchRunArtifacts(
+          any,
+          perPage: anyNamed('perPage'),
+          page: anyNamed('page'),
+        )).thenErrorWith();
+
+        final result = adapter.fetchBuilds(testDataGenerator.jobName);
+
+        expect(result, throwsStateError);
+      },
+    );
+
+    test(
+      ".fetchBuilds() throws a StateError if paginating through coverage artifacts fails",
```

Comment:
Should we add the following test cases: 
- .fetchBuilds() maps the coverage value to `null` if an artifact with the specified name does not exist 
- .fetchBuilds() maps the coverage value to `null` if an artifact archive does not contain the coverage summary json

And the same for .fetchBuildsAfter()?